File: app/realtime/types/yjs_persistence_batch.go
package realtimetypes

import (
	"errors"

	"github.com/google/uuid"
)

type YjsPersistenceBatch struct {
	PersistenceBatchId uuid.UUID
	OriginConnectionId *uuid.UUID
	Payload            []byte
}

func (b *YjsPersistenceBatch) UnmarshalBytes(payload []byte) error {
	*b = YjsPersistenceBatch{}

	// [persistenceBatchId:16][originConnectionId:16, zero UUID when mixed][raw Yjs update:n]
	if len(payload) <= 32 {
		return errors.New("invalid yjs persistence batch")
	}

	copy(b.PersistenceBatchId[:], payload[:16])
	if b.PersistenceBatchId == uuid.Nil {
		*b = YjsPersistenceBatch{}

		return errors.New("invalid yjs persistence batch")
	}

	var originConnectionId uuid.UUID
	copy(originConnectionId[:], payload[16:32])
	if originConnectionId != uuid.Nil {
		b.OriginConnectionId = &originConnectionId
	}

	b.Payload = payload[32:]

	return nil
}

func (b YjsPersistenceBatch) MarshalBytes() ([]byte, error) {
	if b.PersistenceBatchId == uuid.Nil || len(b.Payload) == 0 {
		return nil, errors.New("invalid yjs persistence batch")
	}

	payload := make([]byte, 32+len(b.Payload))
	copy(payload[:16], b.PersistenceBatchId[:])
	if b.OriginConnectionId != nil {
		copy(payload[16:32], b.OriginConnectionId[:])
	}
	copy(payload[32:], b.Payload)

	return payload, nil
}

File: yjs-worker/src/types/yjs_persistence_batch.ts
import type WebSocket from "ws";

import { convertBytesToUUIDString, convertUUIDToBytes } from "../util/uuid.js";

// YjsPersistenceBatch is the internal Go/worker envelope for one merged raw Yjs update and its idempotency key
export type YjsPersistenceBatch = {
  persistenceBatchId: string;
  originConnectionId: string | null;
  payload: Buffer;
};

// InFlightYjsPersistenceBatch is one merged update awaiting its durable Go/PostgreSQL acknowledgement
export type InFlightYjsPersistenceBatch = {
  persistenceBatchId: string;
  originConnectionId: string | null;
  payload: Buffer;
  webSocket: WebSocket;
  connectionId: string;
  connectorChannelId: number;
  updateCount: number;
};

export function createYjsPersistenceBatch(
  persistenceBatchId: string,
  originConnectionId: string | null,
  payload: Buffer
): Buffer {
  // [persistenceBatchId:16][originConnectionId:16, zero UUID when mixed][raw Yjs update:n]
  const batchPayload = Buffer.alloc(32 + payload.length);

  convertUUIDToBytes(persistenceBatchId).copy(batchPayload, 0);
  if (originConnectionId !== null) {
    convertUUIDToBytes(originConnectionId).copy(batchPayload, 16);
  }
  payload.copy(batchPayload, 32);

  return batchPayload;
}

export function parseYjsPersistenceBatch(
  payload: Buffer
): YjsPersistenceBatch | null {
  if (payload.length <= 32) {
    return null;
  }

  const persistenceBatchId = convertBytesToUUIDString(payload.subarray(0, 16));
  const originConnectionId = convertBytesToUUIDString(payload.subarray(16, 32));
  if (persistenceBatchId === null) {
    return null;
  }

  return {
    persistenceBatchId,
    originConnectionId,
    payload: payload.subarray(32),
  };
}
